pkg/agent/agent_loop_test.go

```diff
@@ -0,0 +1,313 @@
+		}
+	}
+	return result
+}
+
+type mockTextPart struct {
+	gollm.Part
+	text string
+}
+
+func (p *mockTextPart) AsText() (string, bool) {
+	return p.text, true
+}
+
+func (p *mockTextPart) AsFunctionCalls() ([]gollm.FunctionCall, bool) {
+	return nil, false
+}
+
+type mockFunctionCallPart struct {
+	gollm.Part
+	calls []gollm.FunctionCall
+}
+
+func (p *mockFunctionCallPart) AsText() (string, bool) {
+	return "", false
+}
+
+func (p *mockFunctionCallPart) AsFunctionCalls() ([]gollm.FunctionCall, bool) {
+	return p.calls, true
+}
+
+// newMockChatResponse creates a gollm.ChatResponseIterator from a list of parts.
+func newMockChatResponse(parts ...gollm.Part) gollm.ChatResponseIterator {
+	return func(yield func(gollm.ChatResponse, error) bool) {
+		response := &mockChatResponse{
+			candidates: []gollm.Candidate{
+				&mockCandidate{
+					parts: parts,
+				},
+			},
+		}
+		yield(response, nil)
+	}
+}
+
+func TestAgentRun_LLMTextResponse(t *testing.T) {
+	t.Parallel()
+	mockChat := &mockChat{
+		sendStreamingFunc: func(ctx context.Context, contents ...any) (gollm.ChatResponseIterator, error) {
+			return newMockChatResponse(&mockTextPart{text: "hello from the mock"}), nil
+		},
+	}
+	a := newTestAgent(mockChat)
+	a.RunOnce = true
+
+	err := a.Run(context.Background(), "test query")
+	if err != nil {
+		t.Fatalf("Run() error = %v", err)
+	}
+
+	// Expect a text response from the model
+	timeout := time.After(1 * time.Second)
+	var receivedResponse bool
+	for !receivedResponse {
+		select {
+		case msg := <-a.Output:
+			m := msg.(*api.Message)
+			if m.Type == api.MessageTypeText && m.Source == api.MessageSourceModel {
+				if strings.Contains(m.Payload.(string), "hello from the mock") {
+					receivedResponse = true
+				}
+			}
+		case <-timeout:
+			t.Fatal("timed out waiting for model response")
+		}
+	}
+
+	for range a.Output {
+		if a.AgentState() != api.AgentStateExited {
+			t.Errorf("Expected agent state to be Exited, but got %s", a.AgentState())
+		}
+	}
+}
+
+func TestAgentRun_LLMErrorResponse(t *testing.T) {
+	t.Parallel()
+	expectedErr := errors.New("mock LLM error")
+	mockChat := &mockChat{
+		sendStreamingFunc: func(ctx context.Context, contents ...any) (gollm.ChatResponseIterator, error) {
+			return nil, expectedErr
+		},
+	}
+	a := newTestAgent(mockChat)
+	a.RunOnce = true
+
+	err := a.Run(context.Background(), "test query")
+	if err != nil {
+		t.Fatalf("Run() error = %v", err)
+	}
+
+	// Expect an error message from the agent
+	timeout := time.After(2 * time.Second)
+	var receivedResponse bool
+	for !receivedResponse {
+		select {
+		case msg := <-a.Output:
+			m := msg.(*api.Message)
+			if m.Type == api.MessageTypeError && m.Source == api.MessageSourceAgent {
+				if strings.Contains(m.Payload.(string), expectedErr.Error()) {
+					receivedResponse = true
+				}
+			}
+		case <-timeout:
+			t.Fatal("timed out waiting for error message")
+		}
+	}
+
+	for range a.Output {
+		if a.AgentState() != api.AgentStateExited {
```

Comment:
Should this be out of the for loop ? to ensure agent's final state to be in the exited ?